userspace/libscap/scap_procs.c

```diff
@@ -1002,7 +1002,20 @@ static int32_t _scap_proc_scan_proc_dir_impl(scap_t* handle, char* procdirname,
 		if(res != SCAP_SUCCESS)
 		{
 			snprintf(error, SCAP_LASTERR_SIZE, "cannot add procs tid = %"PRIu64", parenttid = %"PRIi32", dirname = %s, error=%s", tid, parenttid, procdirname, add_error);
-			break;
+			//
+			// Bug 1897: Agent fails to initialize sinsp_worker: error creating the process list
+			//           When parameter suppress is true, report in the log, but do not process 
+			//           this file any further, because it is no longer in the process list.  
+			//           This is not considered a failure.
+			//
+			if(suppress)
+			{
+				res = SCAP_SUCCESS;
```

Comment:
MAJOR: no, we cannot add this process and ignore the errors here. If its /proc entries disappeared halfway through, it means whatever we collected so far is out of date (the thread has just exited). By the time we set up everything, the pid may have been reused by something completely different and since we do have a small race at startup wrt the proc scan (another long-standing bug in this code), we may well use the wrong thread for the event:

1. scan pid=1234, it exits halfway through
2. keep scanning /proc
3. pid=1234 is reused (driver isn't open yet, we don't get the clone event)
4. finish scanning /proc
5. open driver
6. get event for pid=1234, attach it to the one we found in (1), not the one created in (3)

When a /proc lookup fails (while scanning the whole directory, not while looking up a single tid), we should drop this thread/process completely. We'll fill the gap later, when the first event for that process arrives.

It's been a while since I looked at this code in detail, but can we return SCAP_FAILURE from the function that reads a single thread's data from /proc and handle it in the caller? There's a lot of places this can fail (we're reading many /proc files) and we only care about complete success (everything else should cause the thread to be discarded).